src/System.IO.Pipelines/tests/PipePoolTests.cs

```diff
@@ -266,6 +267,114 @@ public async Task OnReaderCompletedCalledAfterBlocksReturned()
             pipe.Reader.Complete();
         }
 
+        [Fact]
+        public async Task WritesToArrayPoolByDefault()
+        {
+            var pipe = new Pipe();
+            pipe.Writer.WriteEmpty(10);
+            await pipe.Writer.FlushAsync();
+            pipe.Writer.Complete();
+
+            var result = await pipe.Reader.ReadAsync();
+            Assert.Equal(10, result.Buffer.Length);
+
+            SequenceMarshal.TryGetReadOnlySequenceSegment(
+               result.Buffer,
+               out ReadOnlySequenceSegment<byte> start,
+               out int startIndex,
+               out ReadOnlySequenceSegment<byte> end,
+               out int endIndex);
+
+            var startSegment = (BufferSegment)start;
+            var endSegment = (BufferSegment)end;
+
+            Assert.Same(startSegment, endSegment);
+            Assert.NotNull(startSegment.Memory);
+            Assert.IsType<byte[]>(startSegment.MemoryOwner);
+
+            pipe.Reader.AdvanceTo(result.Buffer.End);
+            pipe.Reader.Complete();
+        }
+
+        [Fact]
+        public async Task GetMemoryOverMaxPoolSizeAllocatesArray()
+        {
+            using (var pool = new DisposeTrackingBufferPool())
+            {
+                var pipe = new Pipe(new PipeOptions(pool: pool));
+
+                // Allocate 5 KB
+                pipe.Writer.WriteEmpty(5 * 1024);
+                await pipe.Writer.FlushAsync();
+                pipe.Writer.Complete();
+
+                Assert.Equal(0, pool.CurrentlyRentedBlocks);
+
+                var result = await pipe.Reader.ReadAsync();
+                Assert.Equal(5 * 1024, result.Buffer.Length);
+
+                SequenceMarshal.TryGetReadOnlySequenceSegment(
+                   result.Buffer,
+                   out ReadOnlySequenceSegment<byte> start,
+                   out int startIndex,
+                   out ReadOnlySequenceSegment<byte> end,
+                   out int endIndex);
+
+                var startSegment = (BufferSegment)start;
+                var endSegment = (BufferSegment)end;
+
+                Assert.Same(startSegment, endSegment);
+
+                // Null owner implies that the buffer is allocated and wasn't rented from the pool
+                Assert.Null(startSegment.MemoryOwner);
```

Comment:
Interesting.  So this is now a visible change in behavior?  Will it matter to anyone?

Reply:
The property is internal for testing

Reply:
Ah, perfect.